luigi/contrib/snowflake.py

```diff
@@ -0,0 +1,411 @@
+import abc
+import logging
+
+import luigi
+from luigi.contrib import rdbms
+from luigi.contrib import redshift
+from luigi.contrib import postgres
+
+logger = logging.getLogger('luigi-interface')
+
+# Snowflake's library do not have constant for their error code.
+# We define our own
+SNOWFLAKE_TABLE_EXISTS_ERROR = 2002
+SNOWFLAKE_GENERIC_PROGRAMMING_ERROR = 2003
+
+try:
+    import snowflake
+    import snowflake.connector
+except ImportError:
+    logger.warning("Loading snowflake module without snowflake-connector-python installed. "
+                   "Will crash at runtime if snowflake functionality is used.")
+
+
+class _SettingsMixins():
+    """
+    Generic mixins that defines how to fetch the database and warehouse
+    configuration for a given snowflake integration.
+    """
+    @property
+    def configuration_section(self):
+        """
+        Override to change the configuration section used
+        to obtain default credentials.
+        """
+        return 'snowflake'
+
+    @property
+    def warehouse(self):
+        """
+        Fetch the warehouse configuration from the luigi configuration file under:
+
+        [snowflake]
+        warehouse=value
+        """
+        return luigi.configuration.get_config().get(self.configuration_section, 'warehouse', default=None)
+
+
+class _CredentialsMixin(redshift._CredentialsMixin):
```

Comment:
Snowflake's website suggests multiple cloud providers are available. Your implementation auth is specific for AWS. Should AWS be an option which happens to be the default?

Reply:
Yes, you are right, it supports multiple cloud providers. We've only been using AWS so I took the code from the Redshift contrib. I like the idea of the option, I'll see what I can do.

Reply:
I am having a problem and I'd like to have some feedback. Luigi is able to hydrate a property for a given class if that property exists during the initialization of the class.

Given this use case:

```python
MyMixin(object):
    @property
    def aws_access_key_id(self):
        """
        Override to return the key id.
        """
        return luigi.configuration.get_config().get(self.configuration_section, 'aws_access_key_id')

    @property
    def aws_secret_access_key(self):
        """
        Override to return the secret access key.
        """
        return luigi.configuration.get_config().get(self.configuration_section, 'aws_secret_access_key')


MyTask(rdms.CopyTable, MyMixins):
    pass
```

If there is an environment variable named `aws_secret_access_key`, then calling `MyTask.aws_secret_access_key` will yield that value, else it will execute the function defined in the mixin.

My concern with your ask is that we would need to have one Mixin that contains all the different properties. The ones from AWS and the ones from Azure. 

If we do it differently, then hydrating the properties using the environment variable isn't supported anymore and that flow breaks. 

I've added a new commit on this pull request named "--wip--" that illustrate what I'm trying to go for but the specs using "TestInternalCredentials" are obviously failing, because it can't hydrate the object since we can only tell which credentials to use after the class has already been initialized.

Let me know if the problem I'm raising is unclear, I'd need some guidance to solve the issue!
Thanks!